components/engine/src/rocks/sst.rs

```diff
@@ -0,0 +1,181 @@
+            io_options.set_env(env.clone());
+        }
+        io_options.compression(get_fastest_supported_compression_type());
+        // in rocksdb 5.5.1, SstFileWriter will try to use bottommost_compression and
+        // compression_per_level first, so to make sure our specified compression type
+        // being used, we must set them empty or disabled.
+        io_options.compression_per_level(&[]);
+        io_options.bottommost_compression(DBCompressionType::Disable);
+        let mut writer = SstFileWriter::new(EnvOptions::new(), io_options);
+        writer.open(path)?;
+        Ok(SstWriter { writer, env })
+    }
+}
+
+/// SstWriter is used to create sst files that can be added to database later.
+pub struct SstWriter {
+    writer: SstFileWriter,
+    env: Option<Arc<Env>>,
+}
+
+impl SstWriter {
+    /// Add key, value to currently opened file
+    /// REQUIRES: key is after any previously added key according to comparator.
+    pub fn put(&mut self, key: &[u8], val: &[u8]) -> Result<(), String> {
+        self.writer.put(key, val)
+    }
+
+    /// Add a deletion key to currently opened file
+    /// REQUIRES: key is after any previously added key according to comparator.
+    pub fn delete(&mut self, key: &[u8]) -> Result<(), String> {
+        self.writer.delete(key)
+    }
+
+    /// Return the current file size.
+    pub fn file_size(&mut self) -> u64 {
+        self.writer.file_size()
+    }
+
+    /// Finalize writing to sst file and close file.
+    pub fn finish(mut self) -> Result<ExternalSstFileInfo, String> {
+        self.writer.finish()
+    }
+
+    /// Finalize writing to sst file and read the contents into the buffer.
+    pub fn finish_into(mut self, buf: &mut Vec<u8>) -> Result<ExternalSstFileInfo, String> {
+        use std::io::Read;
+        if let Some(env) = self.env.take() {
+            let sst_info = self.writer.finish()?;
+            let p = sst_info.file_path();
+            let path = p.as_os_str().to_str().ok_or_else(|| {
+                format!(
+                    "failed to sequential file bad path {}",
+                    sst_info.file_path().display()
+                )
+            })?;
+            let mut seq_file = env.new_sequential_file(path, EnvOptions::new())?;
+            let len = seq_file
+                .read_to_end(buf)
+                .map_err(|e| format!("failed to read sequential file {:?}", e))?;
+            if len as u64 != sst_info.file_size() {
+                Err(format!(
+                    "failed to read sequential file inconsistent length {} != {}",
+                    len,
+                    sst_info.file_size()
+                ))
+            } else {
+                Ok(sst_info)
+            }
+        } else {
+            Err(format!("failed to read sequential file no env provided"))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::rocks::util;
+    use tempfile::Builder;
+
+    #[test]
+    fn test_somke() {
```

Comment:
What is `somke`?

Reply:
smoke 🤣

Reply:
what is smoke?

Reply:
It's smoke testing. https://en.wikipedia.org/wiki/Smoke_testing_(software)

Reply:
<https://en.wikipedia.org/wiki/Smoke_testing_(software)>